### PR TITLE
Add GraphQL queries for viewing deadletter reports

### PR DESCRIPTION
**Pull request by manu-govind**
DASB-433

### CODE REVIEW

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/ReportDeadLetter.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    var contentType : String? = null,
+
+    @GraphQLDescription("Message id this report belongs to; set to null if not applicable")
+    var messageId: String? = null,
+
+    @GraphQLDescription("Status this report is indicating, such as success or failure")
+    var status : String? = null,
+
+    @GraphQLDescription("Content of the report.  If the report is JSON then the content will be shown as JSON.  Otherwise, the content is a base64 encoded string.")
+    var content: SchemaDefinition? = null,
```
Comment (mkrystof) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/ReportDeadLetter.kt:48]: I think this needs to be a string or an Any type to handle the case where the content is malformed, right?

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import com.azure.cosmos.models.CosmosQueryRequestOptions
+import gov.cdc.ocio.processingstatusapi.models.ReportDeadLetter
+import gov.cdc.ocio.processingstatusapi.models.dao.ReportDao
+import java.text.SimpleDateFormat
+import java.time.ZoneOffset
+import java.util.*
+
+class ReportDeadLetterLoader : CosmosDeadLetterLoader() {
+
+    fun getByUploadId(uploadId: String): List<ReportDeadLetter> {
```
Comment (mkrystof) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt:12]: Please add function and class headers to all files.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                "and r.timestamp<=$endDateParam" + if (dataStreamRoute!=null) " and r.dataStreamRoute= '$dataStreamRoute'" else ""
+
+         val reportItems = reportsDeadLetterContainer.queryItems(
+            reportsSqlQuery, CosmosQueryRequestOptions(),
+            Int::class.java
+        )
+        var count = 0
+        if (reportItems.iterator().hasNext()) {
+             count = reportItems.iterator().next()
+            println("Count of records: $count")
```
Comment (mkrystof) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt:64]: Please replace all println calls with logger calls.  Plenty of examples in code for using klogger.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+class ReportDeadLetterQueryService : Query {
+
+    @GraphQLDescription("Return all the dead-letter reports associated with the provided uploadId")
+    @Suppress("unused")
+    fun getDeadLetterReportsByUploadId(uploadId: String) = ReportDeadLetterLoader().getByUploadId(uploadId)
+
+    @GraphQLDescription("Return all the dead-letter reports associated with the provided datastreamId, datastreamroute and timestamp date range")
+    @Suppress("unused")
+    fun getDeadLetterReportsByDataStream(dataStreamId: String, dataStreamRoute:String, startDate:String, endDate:String)
```
Comment (mkrystof) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt:19]: Please allow for either start/end date OR days interval.  We have a pattern for this in most of our queries to where you can either supply a start date and optional end date or simply the number of days.  You should be able to find lots of examples for this.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+              = ReportDeadLetterLoader().getByDataStreamByDateRange(dataStreamId,dataStreamRoute,startDate,endDate)
+
+    @GraphQLDescription("Return count of dead-letter reports associated with the provided datastreamId, (optional) datastreamroute and timestamp date range")
+    @Suppress("unused")
+    fun getDeadLetterReportsCountByDataStream(dataStreamId: String, dataStreamRoute:String?, startDate:String, endDate:String)
```
Comment (mkrystof) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt:24]: Same comment as above with days interval.

**Review by mkrystof — CHANGES_REQUESTED** [review]

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt (hunk 2 above)
Comment (manu-govind) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt:20]: Done. Please review

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt (hunk 3 above)
Comment (manu-govind) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/loaders/ReportDeadLetterLoader.kt]: Done. Please review

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt (hunk 4 above)
Comment (manu-govind) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt]: Done. Please review

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt (hunk 5 above)
Comment (manu-govind) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/queries/ReportDeadLetterQueryService.kt]: Done. Please review

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/ReportDeadLetter.kt (hunk 1 above)
Comment (manu-govind) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/ReportDeadLetter.kt:48]: Needs to discuss with you on this one. The Report data class also has it of type SchemaDefinition. When Any is given its throwing a run time exception. Hence I've not fixed this one and will make changes once I discuss with you.

Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/ReportDeadLetter.kt:48]: Let's address this one when we add more validation for the report schemas, which will be assigned this current sprint anyway.

**Review by mkrystof — APPROVED** [review]
lgtm

**Pull request merged**